### PR TITLE
strip http2 pseudo-header field from request before http outbound call

### PR DESCRIPTION
**Pull request by AllenLuUber**
- [x] Description and context for reviewers: one partner, one stranger

In case grpc inbound  http outbound, and the request contains http2
specific pseudo-header fields, http outbound call will fail. 

Making the fix on authority field based on the [RFC](https://tools.ietf.org/html/rfc7540#section-8.1.2.3):
An intermediary that converts an HTTP/2 request to HTTP/1.1 MUST
create a Host header field if one is not present in a request by
copying the value of the ":authority" pseudo-header field.

For other pseduo-header fields, http transport will strip those before
forwarding to a http outbound.

The added test will fail before the fix:

```
--- FAIL: TestGRPCInHTTPOut (0.00s)
    outbound_test.go:676:
        	Error Trace:	outbound_test.go:676
        	Error:      	Received unexpected error:
        	            	code:unknown message:unknown error from http client: Post "http://127.0.0.1:63722": net/http: invalid header field name "Rpc-Header-:path"
        	Test:       	TestGRPCInHTTPOut
```
- [x] Entry in CHANGELOG.md
Fixed an issue when sending requests with http2 pseduo-header field to http outbound.

### CODE REVIEW

File: transport/http/outbound.go [hunk 1]
```diff
@@ -53,7 +53,18 @@ var (
 	_ introspection.IntrospectableOutbound = (*Outbound)(nil)
 )
 
-var defaultURLTemplate, _ = url.Parse("http://localhost")
+const http2AuthorityPseudoHeader = ":authority"
+
+var (
+	defaultURLTemplate, _ = url.Parse("http://localhost")
+	// from https://tools.ietf.org/html/rfc7540#section-8.1.2.3
+	http2PseudoHeaders = []string{
```
Comment (jaricftw) [inline transport/http/outbound.go:61]: is this exhaustive? or rather, any psudo header started with `:` is un-parsable/invalid in HTTP/1 right?

File: transport/http/outbound.go [hunk 2]
```diff
@@ -339,7 +349,27 @@ func (o *Outbound) getPeerForRequest(ctx context.Context, treq *transport.Reques
...
+		return nil, err
+	}
+	// grpc in, http out case
+	if treq.Transport == "grpc" {
+		// from https://tools.ietf.org/html/rfc7540#section-8.1.2.3
+		// For ":authority", ...
+		// An intermediary that converts an HTTP/2 request to HTTP/1.1 MUST
+		// create a Host header field if one is not present in a request by
+		// copying the value of the ":authority" pseudo-header field.
+		if v, ok := treq.Headers.Get(http2AuthorityPseudoHeader); ok {
```
Comment (jaricftw) [inline transport/http/outbound.go:363]: the spec says creating the header header if one if not present, should we check for existence?

File: transport/http/outbound.go [hunk 3]
```diff
@@ -339,7 +349,27 @@ func (o *Outbound) getPeerForRequest(ctx context.Context, treq *transport.Reques
...
+			hreq.Host = v
+		}
+		// strip all http2 pseudo-header fields
+		for _, k := range http2PseudoHeaders {
```
Comment (jaricftw) [inline transport/http/outbound.go:367]: hmm, is stripping all the psudo headers the right way? I don't there, is there any other translations needed. I haven't looked in details, but for example, do we need to translate `:path` to http/1 URL? similarly for others. 

I am interested in a general translation protocol between HTTP/2 (gRPC) to HTTP/1 here, if we allow it.

File: transport/http/outbound.go (hunk 1 above)
Comment (AllenLuUber) [reply transport/http/outbound.go:61]: This is exhaustive from the RFC. My intention here is strip those http2 specific ones and have a followup PR that fails fast on other invalid ones starting with `:`

File: transport/http/outbound.go (hunk 2 above)
Comment (AllenLuUber) [reply transport/http/outbound.go:363]: ahh good call. I will make the change.

File: transport/http/outbound.go (hunk 3 above)
Comment (AllenLuUber) [reply transport/http/outbound.go:367]: I thought about the same thing. From RFC, it only mentioned the behavior for `:authority` specifically and other pseudo-header field's behavior were left undefined.

File: transport/http/outbound.go [hunk 4]
```diff
@@ -339,7 +349,27 @@ func (o *Outbound) getPeerForRequest(ctx context.Context, treq *transport.Reques
 
 func (o *Outbound) createRequest(treq *transport.Request) (*http.Request, error) {
 	newURL := *o.urlTemplate
-	return http.NewRequest("POST", newURL.String(), treq.Body)
+	hreq, err := http.NewRequest("POST", newURL.String(), treq.Body)
+	if err != nil {
+		return nil, err
+	}
+	// grpc in, http out case
+	if treq.Transport == "grpc" {
```
Comment (jronak) [inline transport/http/outbound.go:357]: requests reaching http outbound typically have transport as `http`, are we sure that `grpc` would be the transport name in grpc-in cases?

File: transport/http/outbound.go [hunk 5]
```diff
@@ -339,7 +349,27 @@ func (o *Outbound) getPeerForRequest(ctx context.Context, treq *transport.Reques
...
+		// create a Host header field if one is not present in a request by
+		// copying the value of the ":authority" pseudo-header field.
+		if v, ok := treq.Headers.Get(http2AuthorityPseudoHeader); ok && hreq.Host == "" {
+			hreq.Host = v
+		}
+		// strip all http2 pseudo-header fields
+		for _, k := range http2PseudoHeaders {
+			treq.Headers.Del(k)
+		}
+	}
```
Comment (Dogild) [inline transport/http/outbound.go:370]: What do you think about putting this logic into a function `stripHTTP2HeadersIfNeeded()`

**Comment by Dogild** [comment]
Closing as duplicate of https://github.com/yarpc/yarpc-go/pull/2086

**Pull request closed without merging**